### PR TITLE
Convert to using dockerode module, and using cryptoworkbench/solana-amman:* images

### LINKED ISSUE #127: start validator with --clone to sync metaplex program
we think this might work too

--clone - 

related to #63

### LINKED ISSUE #141: Update Docker images to 1.10.*

### PR DESCRIPTION
**Pull request by SvenDowideit**
* [x] explicitly pull the image (it looks like it isn't done explicitly, when i made the PR, i had just built them locally)
   * [x] need to make the 'create container' wait til the 'pull' finished - I thought that was the point of the chaining, but it seems not.
* [x] convert stop container
* [x] separate into start container, start validator, stop validator, remove container
* [x] figure out what to show in the logs UI
   * [x] show the dockerode output for each button
   * [x] show `docker exec -t validator amman log` output when running, which puts that into a ring buffer that the IPC sends?
* [x] ammanrc.json `docker copy` and use on start
   * [x] show it works manually
   * [x] use `exec('docker copy`...)
* [x] make sure we can remove the use of the docker cli
* [x] figure out a `docker exec solana-test-validator amman logs` - the problem is that it there's not non-streaming way to do this, so it needs to be done differently - see https://github.com/apocas/dockerode#streams-goodness
* [x] use toasts to show "pulling", "starting container", "starting validator", etc in-progress
* [x] fix disabled / enabled button styling without forcing the dev to build that every time.

Defered to the next PR

* #165 - WORK OUT the validator cli params? (probably punt to later)
* #166 - get the list of image tags available from docker hub, and figure outhow to "prefer" one
* #167
* #168
* #169
* #170
* #171
* #175

closes #141 
closes #127

### CODE REVIEW

**Comment by SvenDowideit** [comment]
mmm, so in the `npm run dev` console, you can see the `dockerode.pull` handler showing the pull progress, and then create the container - and if you wait, you see the create button get disabled, and the start button enable.

but because the handler is going after the promise returns, the toast completes.

I'd much rather have a promise that waits til the container is created, but i suspect I actually need to implement a manually controlled `toast` that gets updated from the pull handler

**Comment by SvenDowideit** [comment]
image

**Comment by SvenDowideit** [comment]
Okay!, @nathanleclaire can you please review this - I would like to merge it asap so I can use it for the tokens work.

File: src/renderer/nav/Validator.tsx [hunk 1]
```diff
@@ -1,37 +1,95 @@
...
+
+    // if (validator.status === NetStatus.Running) {
+    window.electron.ipcRenderer.validatorLogs({
+      filter: filterRef.current.value || '',
+      net: validator.net,
+    });
+    // }
   }, 5000);
 
+  useEffect(() => {
```
Comment (nathanleclaire) [inline src/renderer/nav/Validator.tsx:74]: should work if you write the function something like this:

```typescript
useEffect(() => {
  const fn = async() => {
    try {
      const tags = await window.promiseIpc.send(
        'DOCKER-GetImageTags', 
        validatorImageName
      );
      setValidatorImageTags(tags);
    } catch (e) {
      logger.error(e);
    }
  };
  fn();
}, [validatorImageName]);
```

File: src/main/ipc/docker.ts [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+    // By default Amman will pull the account data from the accountsCluster (can be overridden on a per account basis)
+    accountsCluster: 'https://api.metaplex.solana.com',
+    accounts: [
+      {
+        label: 'Token Metadata Program',
+        accountId: 'metaqbxxUerdq28cj1RbAWkYQm3ybzjb6a8bt518x1s',
+        // marking executable as true will cause Amman to pull the executable data account as well automatically
+        executable: true,
+      },
+      {
```
[the remark this answers is not in the document]
Comment (nathanleclaire) [reply src/main/ipc/docker.ts:29]: Maybe add a comment that these are like verification accounts

File: src/main/ipc/docker.ts [hunk 3]
```diff
@@ -0,0 +1,356 @@
...
+
+declare type IpcEvent = IpcRendererEvent & IpcMainEvent;
+
+const DOCKER_PATH =
+  process.platform === 'darwin' ? '/usr/local/bin/docker' : 'docker';
+
+const ammanrc = {
+  validator: {
+    // By default Amman will pull the account data from the accountsCluster (can be overridden on a per account basis)
+    accountsCluster: 'https://api.metaplex.solana.com',
```
Comment (SvenDowideit) [inline src/main/ipc/docker.ts:21]: NO? use normal api...

File: src/main/ipc/docker.ts [hunk 4]
```diff
@@ -0,0 +1,356 @@
...
+        accountId: 'SvencSSSMANWBYyqFhMur1r7a3N6WAA1TEQNSEcRBnz',
+      },
+      {
+        label: 'Random other account',
+        accountId: '4VLgNs1jXgdciSidxcaLKfrR9WjATkj6vmTm5yCwNwui',
+        // By default executable is false and is not required to be in the config
+        // executable: false,
+
+        // Providing a cluster here will override the accountsCluster field
+        cluster: 'https://metaplex.devnet.rpcpool.com',
```
Comment (SvenDowideit) [inline src/main/ipc/docker.ts:40]: same here, use normal solana

File: src/main/ipc/docker.ts [hunk 5]
```diff
@@ -0,0 +1,356 @@
...
+      },
+    ],
+    killRunningValidators: true,
+    jsonRpcUrl: 'http://0.0.0.0:8899',
+    websocketUrl: '',
```
Comment (nathanleclaire) [inline src/main/ipc/docker.ts:45]: should this be set? 'ws://0.0.0.0:8900`

File: src/main/ipc/docker.ts [hunk 6]
```diff
@@ -0,0 +1,356 @@
...
+
+      return ['v1.9.29', 'v1.10.27', 'v1.11.1'];
+    }
+  );
+
+  promiseIpc.on(
+    'DOCKER-GetContainerStatus',
+    (name: unknown, event?: IpcEvent | undefined) => {
+      logger.silly(`main: called DOCKER-GetContainerStatus, ${name}, ${event}`);
+      // log(`main: called DOCKER-GetContainerStatus, ${name}, ${event}`);
```
Comment (nathanleclaire) [inline src/main/ipc/docker.ts:224]: rm extra comment please ?

File: src/main/ipc/docker.ts [hunk 7]
```diff
@@ -0,0 +1,356 @@
...
+    'DOCKER-StopValidatorContainer',
+    (image: unknown, event?: IpcEvent | undefined) => {
+      logger.info(
+        `main: called DOCKER-StopValidatorContainer, ${image}, ${event}`
+      );
+
+      const dockerClient = new Dockerode();
+      const container = dockerClient.getContainer('solana-test-validator');
+      logger.error(`request stop: solana-test-validator`);
+      log(`request stop: solana-test-validator`);
```
Comment (nathanleclaire) [inline src/main/ipc/docker.ts:272]: what's up with these multiple `log` calls?

File: src/renderer/nav/Validator.tsx [hunk 8]
```diff
@@ -1,37 +1,95 @@
...
-      window.electron.ipcRenderer.validatorLogs({
-        filter: filterRef.current.value || '',
-        net: validator.net,
-      });
+    if (validator.net === Net.Localhost) {
+      window.promiseIpc
+        .send('DOCKER-GetContainerStatus', 'solana-test-validator')
+        .then((inspect: any) => {
+          // setValidatorImageTags(tags);
+          // logger.info(`INSPECT: ${JSON.stringify(inspect)}`);
```
Comment (nathanleclaire) [inline src/renderer/nav/Validator.tsx:46]: rm pls

File: src/renderer/nav/Validator.tsx [hunk 9]
```diff
@@ -48,62 +106,162 @@ const Validator = () => {
...
-      )}
+        </div>
+      </ButtonToolbar>
+      <InputGroup size="sm">
+        <FormControl
+          ref={filterRef}
+          placeholder="Filter logs"
+          aria-label="Amount"
+          onKeyDown={debounce(() => {
+            // if (validator.status === NetStatus.Running) {
```
Comment (nathanleclaire) [inline src/renderer/nav/Validator.tsx:251]: rm

File: src/renderer/nav/Validator.tsx [hunk 10]
```diff
@@ -48,62 +106,162 @@ const Validator = () => {
...
+                    window.promiseIpc.send(`DOCKER-StartAmmanValidator`);
+                    logger.info('STARTED AMMAN');
+                    return 'ok';
+                  })
+                  .catch(logger.error);
+              } else {
+                // need to wait til the container has started...
+                logger.info('START AMMAN');
+                // TODO: blocking, no toast
+                // ipcDockerToast('StartAmmanValidator');
```
Comment (nathanleclaire) [inline src/renderer/nav/Validator.tsx:201]: rm

File: src/renderer/nav/Validator.tsx [hunk 11]
```diff
@@ -48,62 +106,162 @@ const Validator = () => {
...
+                containerInspect &&
+                containerInspect.State &&
+                !containerInspect.State.Running
+              ) {
+                ipcDockerToast('StartValidatorContainer')
+                  .then(() => {
+                    logger.info('STARTED CONTAINER');
+                    logger.info('START AMMAN');
+                    // TODO: blocking, no toast
+                    // ipcDockerToast('StartAmmanValidator');
```
Comment (nathanleclaire) [inline src/renderer/nav/Validator.tsx:191]: rm

File: src/renderer/nav/Validator.tsx [hunk 12]
```diff
@@ -48,62 +106,162 @@ const Validator = () => {
...
+                      logger.info(`hooray ${JSON.stringify(info)}`);
+                      return info;
+                    })
+                    .then(() => {
+                      // eslint-disable-next-line promise/catch-or-return, promise/no-nesting
+                      ipcDockerToast('StartValidatorContainer').then(() => {
+                        logger.info('STARTED CONTAINER');
+                        logger.info('START AMMAN');
+                        // TODO: blocking, no toast
+                        // ipcDockerToast('StartAmmanValidator');
```
Comment (nathanleclaire) [inline src/renderer/nav/Validator.tsx:168]: rm

File: src/renderer/nav/Validator.tsx [hunk 13]
```diff
@@ -48,62 +106,162 @@ const Validator = () => {
...
+            disabled={validatorImageTag === ''}
+            onClick={() => {
+              logger.info(`GOGOGO ${JSON.stringify(validatorImageTag)}`);
+
+              if (!containerInspect || !containerInspect.State) {
+                toast.promise(
+                  window.promiseIpc
+                    .send('DOCKER-CreateValidatorContainer', validatorImageTag)
+                    .then((info: any) => {
+                      // setValidatorImageTags(tags);
```
Comment (nathanleclaire) [inline src/renderer/nav/Validator.tsx:158]: rm

File: src/renderer/nav/Validator.tsx [hunk 14]
```diff
@@ -1,37 +1,95 @@
...
+  }, [validatorImageName]);
+
   useEffect(() => {
     // eslint-disable-next-line @typescript-eslint/no-explicit-any
     const listener = (resp: any) => {
       const { method, res } = resp;
       switch (method) {
         case 'validator-logs':
-          setValidatorLogs(res);
+          // logger.info('log: ', res);
```
Comment (nathanleclaire) [inline src/renderer/nav/Validator.tsx:90]: rm

File: src/renderer/nav/Validator.tsx [hunk 15]
```diff
@@ -1,37 +1,95 @@
...
+
+          return inspect;
+        })
+        .catch((inspectError: any) => {
+          setContainerInspect({});
+          logger.silly(inspectError);
+        });
     }
+
+    // if (validator.status === NetStatus.Running) {
```
Comment (nathanleclaire) [inline src/renderer/nav/Validator.tsx:66]: rm

File: src/main/validator.ts [hunk 16]
```diff
@@ -1,132 +1,28 @@
...
-
-  // TODO: doing this out of process might be a better fit
-  const maxBuffer = 104857600; // 100MB
-
-  if (!shell.which(DOCKER_PATH)) {
-    return 'Docker executable not found.';
-  }
-
-  // TODO: and now test if there's a Docker daemon up
+  // const MAX_TAIL_LINES = 10000;
```
Comment (nathanleclaire) [inline src/main/validator.ts:16]: rm

File: src/main/ipc/docker.ts [hunk 17]
```diff
@@ -0,0 +1,356 @@
...
+  const stream = await p;
+  container.modem.demuxStream(stream, logStream, logStream);
+  log('Exiting execAmman');
+  return p;
+}
+
+// Need to import the file and call a function (from the main process) to get the IPC promise to exist.
+export function initDockerPromises() {
+  promiseIpc.on(
+    'DOCKER-GetImageTags',
```
Comment (nathanleclaire) [inline src/main/ipc/docker.ts:208]: can merge this in as-is, but in future, can we stick to existing `skewer-case` convention please?

File: src/main/ipc/docker.ts [hunk 18]
```diff
@@ -0,0 +1,356 @@
...
+    .exec({
+      Cmd: ['/bin/bash', '-c', 'source /root/.bashrc && amman start'],
+      AttachStdin: true,
+      AttachStdout: true,
+    })
+    .then((exec: Dockerode.Exec) => {
+      log('start amman exec created');
+
+      return exec.start({
+        // hijack: true,
```
Comment (nathanleclaire) [inline src/main/ipc/docker.ts:192]: rm

File: src/main/ipc/docker.ts [hunk 19]
```diff
@@ -0,0 +1,356 @@
...
+
+async function createContainer(image: string) {
+  const dockerClient = new Dockerode();
+  log(`Pull ${image}`);
+  const pullStream = await dockerClient.pull(image);
+  // logger.info(`----------- pullStream ${JSON.stringify(pullStream)}`);
+
+  function onPullProgress(_event: any) {
+    // ...
+    // logger.info(`onPullProgress: ${JSON.stringify(_event)}`);
```
Comment (nathanleclaire) [inline src/main/ipc/docker.ts:111]: rm

File: src/main/ipc/docker.ts [hunk 20]
```diff
@@ -0,0 +1,356 @@
...
+  }
+
+  return '';
+};
+
+async function createContainer(image: string) {
+  const dockerClient = new Dockerode();
+  log(`Pull ${image}`);
+  const pullStream = await dockerClient.pull(image);
+  // logger.info(`----------- pullStream ${JSON.stringify(pullStream)}`);
```
Comment (nathanleclaire) [inline src/main/ipc/docker.ts:107]: rm

File: src/main/ipc/docker.ts [hunk 21]
```diff
@@ -0,0 +1,356 @@
...
+      },
+      {
+        label: 'Svens debug account',
+        accountId: 'SvencSSSMANWBYyqFhMur1r7a3N6WAA1TEQNSEcRBnz',
+      },
+      {
+        label: 'Random other account',
+        accountId: '4VLgNs1jXgdciSidxcaLKfrR9WjATkj6vmTm5yCwNwui',
+        // By default executable is false and is not required to be in the config
+        // executable: false,
```
Comment (nathanleclaire) [inline src/main/ipc/docker.ts:37]: rm

**Review by nathanleclaire — CHANGES_REQUESTED** [review]

**Comment by SvenDowideit** [comment]
Running with it, will work on the other bits post token

**Pull request merged**